beep-beep/jobs.md

```diff
@@ -0,0 +1,695 @@
+
+For example, in this way, you can configure both the client and server parts to
+work with RabbitMQ.
+
+```yaml
+#
+# RPC is required for tasks dispatching (client)
+#
+rpc:
+  listen: tcp://127.0.0.1:6001
+
+#
+# This section configures the task consumer (server)
+#
+server:
+  command: php consumer.php
+  relay: pipes
+
+#
+# AMQP connection settings (for example, RabbitMQ)
+#
+amqp:
+    addr: amqp://guest:guest@localhost:5672
+
+#
+# In this section, the jobs themselves are configured
+#
+jobs:
+  consume: [ "test" ]   # List of RoadRunner queues that can be processed by 
+                        # the consumer specified in the "server" section.
+  pipelines:
+    test:               # RoadRunner queue identifier
+      driver: amqp      # - Queue driver name
+      queue: test       # - Internal (driver's) queue identifier
+```
+
+- The `rpc` section is responsible for client settings. It is at this address
+  that we will connect, *dispatching tasks* to the queue.
+
+- The `server` section is responsible for configuring the server. Previously, we
+  have already met with its description when setting up the [PHP Worker](/php/worker.md).
+
+- The `amqp` section contains settings for connecting to RabbitMQ. As the name
+  implies, the settings are specific only for the RabbitMQ driver and *may not*
+  be available if you use some other driver.
+
+- And finally, the `jobs` section is responsible for the work of the queues
+  themselves. It contains information on how the RoadRunner should work with
+  connections to drivers, what can be handled by the consumer, and other
+  queue-specific settings.
+
+### Common Configuration
+
+Let's now focus on the common settings of the queue server. In full, it may 
+look like this:
+
+```yaml
+jobs:
+  num_pollers: 64
+  pipeline_size: 100000
+  pool:
+    num_workers: 10
+    max_jobs: 0
+    allocate_timeout: 60s
+    destroy_timeout: 60s
+  consume: [ "queue-name" ]
+  pipelines:
+    queue-name:
+      driver: # "[DRIVER_NAME]"
+      # And driver-specific configuration below...
+```
+
+Above is a complete list of all possible common Jobs settings. Let's now figure
+out what they are responsible for.
+
+- `num_pollers` - TODO
+- `pipeline_size` - TODO
+- `pool.num_workers` - TODO
+- `pool.max_jobs` - TODO
+- `pool.allocate_timeout` - TODO
+- `pool.destroy_timeout` - TODO
+- `consume` - Contains an array of the names of all queues specified in the
+  `"pipelines"` section, which should be processed by the concierge specified in
+   the global `"server"` section (see the [PHP worker's settings](/php/worker.md)).
+- `pipelines` - This section contains a list of all queues declared in the
+  RoadRunner. The key is a unique *queue identifier*, and the value is an object
+  from the settings specific to each driver (we will talk about it later).
+
+### Ephemeral (Memory) Driver
+
+TODO
+
+### AMQP Driver
+
+TODO
+
+### Beanstalk Driver
+
+TODO
+
+### SQS Driver
+
+TODO
+
```

Comment:
```json
---
GLOBAL
  key: api-key
  secret: api-secret
  region: us-west-1
  endpoint: http://localhost:9324
---
LOCAL
      driver: sqs
      prefetch: 1000
      visibility_timeout: 0
      wait_time_seconds: 0
      queue: default
      attributes:
        DelaySeconds: 0
        MaximumMessageSize: 262144
        MessageRetentionPeriod: 345600
        ReceiveMessageWaitTimeSeconds: 0
        VisibilityTimeout: 30
      tags:
        test: "tag"
```
Where:
1. `driver` - driver name.
2. `prefetch` - is the maximum number of messages to return. Amazon SQS never returns more messages than this value (however, fewer messages might be returned). Valid values: 1 to 10. Default: 10. Any number bigger than 10 will be rounded to 10.
2. `visibility_timeout` - The duration (in seconds) that the received messages are hidden from subsequent **retrieve** requests after being retrieved by a ReceiveMessage request. Max: 43200 seconds(12 hours). Default: 0.
3. wait_time_seconds` - The duration (in seconds) for which the call waits for a message to arrive in the queue before returning. If a message is available, the call returns sooner than WaitTimeSeconds. If no messages are available and the wait time expires, the call returns successfully with an empty list of messages. Default: 5.
4. `queue` - sqs queue name. Valid values: alphanumeric characters, hyphens (-), and underscores (_). Default: default.
5. `attributes` - List of all attributes are here: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SetQueueAttributes.html. Default: none.
6. `tags` - just queue tags. Can be set and used by the external queue readers (not RR yet). Default: none.

Notes:
1. It's recommended to wait for a few seconds after the queue is created. Thus RR will sleep for the 2 seconds when SQS queue was declared.
2. SQS will automatically handle redials.
3. Delay can't be more than 15 minutes, thus all values greater than 900 seconds will be rejected with error.